glean-core/rlb/tests/interruptible_shutdown.rs

```diff
@@ -0,0 +1,146 @@
+    use once_cell::sync::Lazy;
+
+    #[allow(non_upper_case_globals)]
+    pub static validation: Lazy<PingType> = Lazy::new(|| {
+        glean::private::PingType::new("validation", true, true, true, true, true, vec![], vec![])
+    });
+}
+
+// Define a fake uploader that reports when and what it uploads.
+#[derive(Debug)]
+struct ReportingUploader {
+    calls: AtomicUsize,
+    sender: Sender<JsonValue>,
+}
+
+impl net::PingUploader for ReportingUploader {
+    fn upload(&self, upload_request: net::PingUploadRequest) -> net::UploadResult {
+        let calls = self.calls.fetch_add(1, Ordering::SeqCst);
+        let body = upload_request.body;
+        let decode = |body: Vec<u8>| {
+            let mut gzip_decoder = GzDecoder::new(&body[..]);
+            let mut s = String::with_capacity(body.len());
+
+            gzip_decoder
+                .read_to_string(&mut s)
+                .ok()
+                .map(|_| &s[..])
+                .or_else(|| std::str::from_utf8(&body).ok())
+                .and_then(|payload| serde_json::from_str(payload).ok())
+                .unwrap()
+        };
+
+        match calls {
+            // First goes through immediately.
+            0 => {
+                self.sender.send(decode(body)).unwrap();
+                net::UploadResult::http_status(200)
+            }
+            // Second SHOULD NEVER SEND
+            1 => panic!("We should shutdown before getting to the second ping"),
+            // Any others ought to be impossible.
+            _ => panic!("Wat."),
+        }
+    }
+}
+
+/// Test scenario: We can interrupt a long glean.upload Wait during shutdown.
+///
+/// The app is initialized, in turn Glean gets initialized without problems.
+/// A custom ping is submitted once, triggering ping throttling.
+/// It is submitted a second time to convince `glean.upload` to be in a `Wait` task.
+/// From this position we ask for Glean to promptly shut down.
+/// We expect this to happen reasonably quickly (within 2s) instead of waiting for the
+/// entire throttling interval _or_ for uploader_shutdown's last-ditch timeout.
+#[test]
+fn interruptible_shutdown() {
+    common::enable_test_logging();
+
+    // Create a custom configuration to use our reporting uploader.
+    let dir = tempfile::tempdir().unwrap();
+    let tmpname = dir.path().to_path_buf();
+    let (tx, rx) = bounded(1);
+
+    let mut cfg = ConfigurationBuilder::new(true, tmpname.clone(), "glean-interruptible-shutdown")
+        .with_server_endpoint("invalid-test-host")
+        .with_use_core_mps(false)
+        .with_uploader(ReportingUploader {
+            calls: AtomicUsize::new(0),
+            sender: tx,
+        })
+        .build();
+    cfg.rate_limit = Some(PingRateLimit {
+        seconds_per_interval: 600, // Needs only be longer than the timeout in `glean::uploader_shutdown()`.
+        pings_per_interval: 1,     // throttle thyself immediately.
+    });
+    common::initialize(cfg);
```

Comment:
while you're here, can you add the `_ = &*pings::validation` dance, while I figure out how to make that less of a problem in the future?

Reply:
Not sure it's needed? Registering the ping will happen immediately before submitting it. We don't even record anything into that ping.

Reply:
Ah fine, I might have been overeager then because I've been doing that so much in my code lately :D